client/utils/src/pubsub.rs

```diff
@@ -0,0 +1,248 @@
+//! - `Registry` — implementation of the subscription/dispatch logic.
+//!
+//! A Registry is implemented by defining the following traits:
+//! - `Subscribe<K>`;
+//! - `Dispatch<M>`;
+//! - `Unsubscribe`.
+//!
+//! As a result of subscription `Hub::subscribe` method returns an instance of `Receiver<Message,
+//! Registry>`. That can be used as a `futures::Stream` to receive the messages.
+//! Upon drop the `Receiver<Message, Registry>` shall unregister itself from the `Hub`.
+
+use std::{
+	collections::HashMap,
+	ops::DerefMut,
+	pin::Pin,
+	sync::{Arc, Weak},
+	task::{Context, Poll},
+};
+
+use ::futures::stream::{FusedStream, Stream};
+use ::parking_lot::Mutex;
+
+use crate::{
+	id_sequence::SeqID,
+	mpsc::{TracingUnboundedReceiver, TracingUnboundedSender},
+};
+
+/// Unsubscribe: unregisters a previously created subscription.
+pub trait Unsubscribe {
+	/// Remove all registrations of the subscriber with ID `subs_id`.
+	fn unsubscribe(&mut self, subs_id: SeqID);
+}
+
+/// Subscribe using a key of type `K`
+pub trait Subscribe<K> {
+	/// Register subscriber with the ID `subs_id` as having interest to the key `K`.
+	fn subscribe(&mut self, subs_key: K, subs_id: SeqID);
+}
+
+/// Dispatch a message of type `M`.
+pub trait Dispatch<M> {
+	/// The type of the that shall be sent through the channel as a result of such dispatch.
+	type Item;
+	/// The type returned by the `dispatch`-method.
+	type Ret;
+
+	/// Dispatch the message of type `M`.
+	///
+	/// The implementation is given an instance of `M` and is supposed to invoke `dispatch` for
+	/// each matching subscriber, with an argument of type `Self::Item` matching that subscriber.
+	///
+	/// Note that this does not have to be of the same type with the item that will be sent through
+	/// to the subscribers. The subscribers will receive a message of type `Self::Item`.
+	fn dispatch<F>(&mut self, message: M, dispatch: F) -> Self::Ret
+	where
+		F: FnMut(&SeqID, Self::Item);
+}
+
+/// A subscription hub.
+///
+/// Does the subscription and dispatch.
+/// The exact subscription and routing behaviour is to be implemented by the Registry (of type `R`).
+/// The Hub under the hood uses the channel defined in `crate::mpsc` module.
+#[derive(Debug)]
+pub struct Hub<M, R> {
+	tracing_key: &'static str,
+	shared: Arc<Mutex<Shared<M, R>>>,
+}
+
+/// The receiving side of the subscription.
+///
+/// The messages are delivered as items of a `futures::Stream`.
+/// Upon drop this receiver unsubscribes itself from the `Hub`.
+#[derive(Debug)]
+pub struct Receiver<M, R>
+where
+	R: Unsubscribe,
+{
+	rx: TracingUnboundedReceiver<M>,
+
+	shared: Weak<Mutex<Shared<M, R>>>,
+	subs_id: SeqID,
+}
+
+#[derive(Debug)]
+struct Shared<M, R> {
+	id_sequence: crate::id_sequence::IDSequence,
+	registry: R,
+	sinks: HashMap<SeqID, TracingUnboundedSender<M>>,
+}
+
+impl<M, R> AsMut<R> for Shared<M, R> {
+	fn as_mut(&mut self) -> &mut R {
+		&mut self.registry
+	}
+}
```

Comment:
This is weird to have here. Especially as you just have this for the `lock_registry` method below.

Below you can just change the return type to `impl DerefMut<Target = &`a mut R> + 'a`

And use https://docs.rs/lock_api/0.4.6/lock_api/struct.MutexGuard.html#method.map to map the value.